File: lua/puzzle5-1.lua
require "utils"

local seeds = io.read()
assert(seeds)
seeds = string.split(seeds, ": ")[2]
assert(seeds)
seeds = string.split(string.strip(seeds))
for i in ipairs(seeds) do seeds[i] = seeds[i] + 0 end
print('seeds', dump(seeds))
io.read()

local maps = {}
for line in io.lines() do
	local map = {}
	map.name = string.split(line)[1]
	assert(map.name)
	map.ranges = {}
	for rngline in io.lines() do
		if rngline == "" then break end
		local range = string.split(rngline)
		assert(#range == 3)
		range.len = table.remove(range) + 0
		range.src = table.remove(range) + 0
		range.dst = table.remove(range) + 0
		table.insert(map.ranges, range)
	end
	table.insert(maps, map)
end

local locations = {}
for _, seed in ipairs(seeds) do
	for _, map in ipairs(maps) do
		for _, range in ipairs(map.ranges) do
			if range.src <= seed and seed < range.src + range.len then
				seed = seed - range.src + range.dst
				break
			end
		end
	end
	table.insert(locations, seed)
end

local min = locations[1]
for _, loc in ipairs(locations) do
	if loc < min then min = loc end
end
print(min)

File: rust/src/bin/puzzle5-1.rs
fn parse_numbers(s: &str) -> Vec<u64> {
    s.split_ascii_whitespace()
        .map(|s| s.parse::<u64>().unwrap())
        .collect()
}

fn read_line() -> Option<String> {
    Some(std::io::stdin().lines().next()?.unwrap())
}

struct Range {
    dst_start: u64,
    src_start: u64,
    src_end: u64,
}

struct Map {
    ranges: Vec<Range>,
}

fn main() {
    let mut seeds = parse_numbers(read_line().unwrap().split_once(':').unwrap().1);
    assert_eq!(read_line().unwrap().len(), 0);

    let mut maps = Vec::new();
    while let Some(line) = read_line() {
        assert!(line.ends_with("map:"));
        let mut ranges = Vec::new();
        for line in std::io::stdin().lines().map(|s| s.unwrap()) {
            if line.is_empty() {
                break;
            }
            let [dst, src, len] = parse_numbers(&line).try_into().unwrap();
            ranges.push(Range {
                dst_start: dst,
                src_start: src,
                src_end: src + len - 1,
            });
        }
        maps.push(Map { ranges });
    }

    for map in &maps {
        for seed in &mut seeds {
            for range in &map.ranges {
                if range.src_start <= *seed && *seed <= range.src_end {
                    *seed = *seed - range.src_start + range.dst_start;
                    break;
                }
            }
        }
    }
    println!("{}", seeds.into_iter().min().unwrap());
}
